Update Sony2Fp tags

postscript-dev opened the pull request:
Changes:
+ Add pretty printing for all tags
+ Add camera selector, to prevent incompatible models using tags
+ Update testing with new values

Source: Exiftool (https://exiftool.org/TagNames/Sony.html)

postscript-dev committed:
Update Sony2Fp makernote tags (tag 9402)

Changes:
+ Add pretty printing for all tags
+ Add camera selector, to prevent incompatible models using tags
+ Update testing with new values

Source: Exiftool (https://exiftool.org/TagNames/Sony.html)

postscript-dev committed:
Merge branch 'Exiv2:main' into update_sony2fp_tags

kevinbackhouse commented on src/makernote_int.cpp:
```diff
@@ -1190,6 +1190,18 @@ namespace Exiv2 {
+
+        std::string model = getExifModel(pRoot);
+        for (auto&& m : models) {
+            if (model.find(m) == 0)
```
It's frustrating that [`starts_with`](https://en.cppreference.com/w/cpp/string/basic_string/starts_with) is a C++20 feature, so I don't think we can use it yet. I would suggest adding a copy of [`startsWith`](https://github.com/Exiv2/exiv2/blob/6798be637f9b36675edcae4504e1dee18091462f/src/epsimage.cpp#L138) to this file, with a TODO comment to replace it with `starts_with` in the future.

postscript-dev replied:
I have added the function.
Thanks.

postscript-dev committed:
Update Sony2Fp makernote tag selector function

In C++20, std::string has a starts_with() function. Add a startsWith()
function now, which can be replaced when the project moves to C++20.

Suggested in https://github.com/Exiv2/exiv2/pull/1777

postscript-dev committed:
Merge branch 'update_sony2fp_tags' of https://github.com/postscript-dev/exiv2 into update_sony2fp_tags

kevinbackhouse commented on src/sonymn_int.cpp:
```diff
@@ -791,21 +792,87 @@ namespace Exiv2 {
+
+            std::string model = pos->toString();
+            for (auto&& m : models) {
+                if (model.find(m) == 0) {
```
Please could you use `startsWith` here too? Sorry for highlighting it last time.

postscript-dev replied:
I have added this.

hassec commented on src/makernote_int.cpp:
```diff
@@ -1190,6 +1198,18 @@ namespace Exiv2 {
+        const char* models[] = { "SLT-", "HV", "ILCA-" };
+
+        std::string model = getExifModel(pRoot);
+        for (auto&& m : models) {
```
```suggestion
        for (auto& m : models) {
```

What do you think about just using `auto&` here ? 
I'm not a fan of using `auto&&` in range based for's unless one actually needs to bind to a rvalue.   
If that isn't necessary, I think it's easier to read and understand `auto&` because `auto&&` just makes people wonder if there is a special thing going on that they didn't yet figure out.

hassec replied:
Actually, you could even do: 
```cpp
for (auto& m : { "SLT-", "HV", "ILCA-" }) {
```

postscript-dev replied:
I have made the change.

hassec commented on src/sonymn_int.cpp:
```diff
@@ -791,21 +792,87 @@ namespace Exiv2 {
+            os << value;
+        else {
+            long val = (value.toLong() & 0x7F);
+            if (val == 0)
```
What do you think about turning this into a switch statement? 
That's usually preferred over `if () else if() ...`, see e.g. [CppCoreGuidelines ES.70](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-switch-if)

postscript-dev replied:
I changed these over. I have other tags still to submit and will use this in future.

postscript-dev committed:
Update Sony2Fp makernote functions

Minor changes using suggestions in #1777:
+ Simplify loop conditions
+ Add startsWith() in SonyMakerNote class
+ Change `if`/`else` to `switch`

kevinbackhouse commented on src/sonymn_int.hpp:
```diff
@@ -71,6 +75,11 @@ namespace Exiv2 {
         static const TagInfo tagInfoFp_[];
         static const TagInfo tagInfoSonyMisc1_[];
         static const TagInfo tagInfo2010e_[];
+
+        // TODO: When moving to C++20, this can be replace with
+        //       std::basic_string<CharT,Traits,Allocator>::starts_with() .
+        //       Suggested in https://github.com/Exiv2/exiv2/pull/1777 .
+        static bool startsWith(const std::string& s, const std::string& start);
```
I wouldn't put this in the header file, since this function is only intended to be used in one source file (`sonymn_int.cpp`). It looks like you need to move the definition of `startWith` to the top of the cpp file. It is currently on line 958, which is after the use on line 867.

postscript-dev replied:
I agree, I tried adding the function in the way that you suggest. The problem was that the `startsWith()` function in `makernote_int.cpp` conflicted when linking - both definitions were in the same namespace.
I attempted adding the definition to `makernote_int.hpp`, but this didn't imediately work and I was trying to avoid adding includes to `sonymn_int.cpp`.

I am open to suggestions.

kevinbackhouse replied:
Got it, sorry about that. Ok, it looks to me like the best plan is to add it to `src/utils.hpp` and `src/utils.cpp`. Please could you remove the copy from `epsimage.cpp` too and get all three files to include "utils.hpp"? Thanks!

postscript-dev replied:
No problem, thanks for the help.
I have made the changes that you suggested.

postscript-dev committed:
Move all `startsWith()` funcs to `Util` namespace

The function checks if a string is in the beginning of another. This
is intended for use until the move to C++20 when
`std::basic_string<>::starts_with()` becomes available. Suggested in
https://github.com/Exiv2/exiv2/pull/1777 .

Separate identical definitions have been moved into a single copy.

postscript-dev commented:
Looking at the failing test, this is a coverage issue and to resolve this would require more test images. Is this something that you want me to look into?

kevinbackhouse commented:
@postscript-dev: The code coverage looks ok to me. As far as I can see, it's all error handling code where it returns early because it has encountered an unexpected value. I think the fuzzer will be able to hit those codepaths fairly easily.
